processing/src/main/java/io/druid/query/groupby/GroupByQueryEngine.java

```diff
@@ -395,11 +498,20 @@ public Row apply(@Nullable Map.Entry<ByteBuffer, Integer> input)
                   ByteBuffer keyBuffer = input.getKey().duplicate();
                   for (int i = 0; i < dimensions.size(); ++i) {
                     final DimensionSelector dimSelector = dimensions.get(i);
-                    final DimExtractionFn fn = dimensionSpecs.get(i).getDimExtractionFn();
+                    final DimensionSpec dimensionSpec = dimensionSpecs.get(i);
+                    final ExtractionFn extractionFn = dimensionSpec.getExtractionFn();
                     final int dimVal = keyBuffer.getInt();
                     if (dimSelector.getValueCardinality() != dimVal) {
-                      if (fn != null) {
-                        theEvent.put(dimNames.get(i), fn.apply(dimSelector.lookupName(dimVal)));
+                      if (extractionFn != null) {
+                        final String value;
+                        // extraction functions are applied at the dimSelector level for time
```

Comment:
not optimizing for time would mean that consecutive (in scan order), but different time values that map to the same resulting string (e.g. `Monday`) would each have to have a different `dimVal` index instead of sharing the same one, because the mapping cannot be used earlier when it actually scans the column (https://github.com/druid-io/druid/pull/1159/files#diff-7d67ab242d530f3d34f752babacd7090L364)

It would also mean that dimSelector.lookupName() would need to support returning long values or Object instead of just String in order to not incur long serialization, which is a much bigger undertaking.